### PR TITLE
Collations Module Integration

### PR DESCRIPTION
**Pull request by king-11**
### Integration of Collation Module into Vitess Engine

- #8991

Adds in the collation module into the Vitess Code which can then be used to replace instances of the `weight_string` function used in the vitess codebase. The major change that this PR makes is with the `NullSafeComparator`

https://github.com/vitessio/vitess/blob/315d90751886b5901672a7278623de7d48aa4ec2/go/vt/vtgate/evalengine/arithmetic.go#L187

which now takes in additional collation info parameter which is used by the `collations.LookupByID` to find out the collation and then use it to collate the `varchar` strings based on their collation set.

### Update Function Calls

So given the `NullSafeComparator` needs access to collation info we intend to provide the information available to it by updating the function calls and trying to fetch charset info from the `result` or `Fields` available at the least distant parent function calls and trickle it down the succeeding calls.

## Related Issue(s)

- #8606 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

**Comment by systay** [comment]
Another general comment - since we need to do planner work to make sure that the engine primitives and evalengine get the right info, what about focusing this PR to make it work well with grouping (`orderedAggregate`)? We can work on the rest of the primitives once we are happy with grouping.

**Comment by vmg** [comment]
@systay: Yes, but do note that the field in `Request.Fields` that contains the numeric identifier for the _collation_ is actually called `charsetnr`, even though it's a _collation identifier_. MySQL is silly like that.

**Comment by vmg** [comment]
> what about focusing this PR to make it work well with grouping (orderedAggregate)? We can work on the rest of the primitives once we are happy with grouping.

This sounds like a great idea. Let's make grouping work perfectly and leave the other callsites for another time. This means that everywhere else we don't pass a collation ID to the comparison function. That should be an easy way to unblock @king-11.

**Comment by king-11** [comment]
@systay @frouioui awaiting your review made the changes and added in tests

File: go/vt/vtgate/engine/ordered_aggregate.go [hunk 1]
```diff
@@ -186,6 +186,17 @@ func (oa *OrderedAggregate) GetTableName() string {
 	return oa.Input.GetTableName()
 }
 
+// GetCollations specifies the collation ID value for columns.
+func (oa *OrderedAggregate) GetCollations() map[int]collations.ID {
```
Comment (systay) [inline go/vt/vtgate/engine/ordered_aggregate.go:190]: this method is only used from this file and it's tests, so I don't think we need to make it public

File: go/vt/vtgate/engine/ordered_aggregate_test.go [hunk 2]
```diff
@@ -1050,3 +1053,137 @@ func TestMultiDistinct(t *testing.T) {
...
+		fields,
+		"a|1",
+		"A|1",
+		"Ǎ|1",
+		"b|2",
+		"c|7",
+		"\xE3\x83\x8F\xE3\x81\xAF|2",
+		"\xE3\x83\x8F\xE3\x83\x8F|1",
+	)
+	fmt.Println(oa.GetCollations())
```
Comment (systay) [inline go/vt/vtgate/engine/ordered_aggregate_test.go:1187]: forgot debug code?

File: go/vt/vtgate/engine/ordered_aggregate_test.go [hunk 3]
```diff
@@ -652,13 +653,15 @@ func TestMerge(t *testing.T) {
 		"1|3|2.8|2|bc",
 	)
 
-	merged, _, err := oa.merge(fields, r.Rows[0], r.Rows[1], nil)
+	// TODO(king-11) make collation aware
```
Comment (systay) [inline go/vt/vtgate/engine/ordered_aggregate_test.go:656]: shouldn't this work if we are adding collations to ordered_aggregate?

File: go/vt/vtgate/evalengine/arithmetic.go [hunk 4]
```diff
@@ -212,6 +223,22 @@ func NullsafeCompare(v1, v2 sqltypes.Value) (int, error) {
...
 		return bytes.Compare(v1.ToBytes(), v2.ToBytes()), nil
 	}
+	if v1.IsText() && v2.IsText() && collationID != collations.Unknown {
+		collation := collations.FromID(collationID)
+		if collation == nil {
+			return 0, UnsupportedCollationError{
+				ID: collationID,
+			}
+		}
+		switch result := collation.Collate(v1.ToBytes(), v2.ToBytes(), false); {
```
Comment (systay) [inline go/vt/vtgate/evalengine/arithmetic.go:233]: we don't really need to switch on this, right? we should be able to just return the value from `collation.Collate` here, I think

File: go/vt/vtgate/vindexes/consistent_lookup.go [hunk 5]
```diff
@@ -314,7 +315,7 @@ func (lu *clCommon) Delete(vcursor VCursor, rowsColValues [][]sqltypes.Value, ks
 func (lu *clCommon) Update(vcursor VCursor, oldValues []sqltypes.Value, ksid []byte, newValues []sqltypes.Value) error {
 	equal := true
 	for i := range oldValues {
-		result, err := evalengine.NullsafeCompare(oldValues[i], newValues[i])
+		result, err := evalengine.NullsafeCompare(oldValues[i], newValues[i], collations.Unknown)
```
Comment (systay) [inline go/vt/vtgate/vindexes/consistent_lookup.go:318]: `TODO`?

**Review by frouioui — APPROVED** [review]
Nothing more than what @systay wrote to add. Otherwise lgtm

File: go/vt/vtgate/evalengine/arithmetic.go (hunk 4 above)
Comment (king-11) [reply go/vt/vtgate/evalengine/arithmetic.go:233]: `Collate` returns a positive or negative or zero value and `NullSafeCompare` needs to return -1,0 or 1.

File: go/vt/vtgate/engine/ordered_aggregate.go (hunk 1 above)
Comment (king-11) [reply go/vt/vtgate/engine/ordered_aggregate.go]: made the change

File: go/vt/vtgate/vindexes/consistent_lookup.go (hunk 5 above)
Comment (king-11) [reply go/vt/vtgate/vindexes/consistent_lookup.go:319]: added `TODO`

File: go/vt/vtgate/engine/ordered_aggregate_test.go (hunk 3 above)
Comment (king-11) [reply go/vt/vtgate/engine/ordered_aggregate_test.go:656]: for this test we don't need to pass in collation I added `TODO` by mistake this works as is supposed to.

Comment (king-11) [reply go/vt/vtgate/engine/ordered_aggregate_test.go]: made the change removed `TODO`

**Comment by king-11** [comment]
@systay made the changes as per your review.

**Comment by shlomi-noach** [comment]
I'm not sure _how_, but this seems to have broken `main`. The following line: 

https://github.com/vitessio/vitess/blob/ab2f0b442e81cc6b5700ffaedc9d8b755898b6b5/go/vt/vtgate/evalengine/arithmetic.go#L227

Does not compile:
```
# vitess.io/vitess/go/vt/vtgate/evalengine
go/vt/vtgate/evalengine/arithmetic.go:227:16: undefined: collations.FromID
```

right now unit tests and some vreplication tests are failing.

**Comment by systay** [comment]
#9131 contains the fix for main :point_up:

**Pull request merged**